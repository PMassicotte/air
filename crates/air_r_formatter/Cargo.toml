
[package]
authors.workspace    = true
categories.workspace = true
description          = "Air R Formatter"
edition.workspace    = true
homepage.workspace   = true
keywords.workspace   = true
license.workspace    = true
name                 = "air_r_formatter"
repository.workspace = true
version              = "0.1.0"

[dependencies]
air_r_syntax    = { workspace = true }
biome_formatter = { workspace = true }
biome_rowan     = { workspace = true }

[lints]
workspace = true
