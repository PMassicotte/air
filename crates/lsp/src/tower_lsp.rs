//
// tower_lsp.rs
//
// Copyright (C) 2022-2024 Posit Software, PBC. All rights reserved.
//
//

#![allow(deprecated)]

use tokio::io::{AsyncRead, AsyncWrite};
use tokio::sync::mpsc::unbounded_channel as tokio_unbounded_channel;
use tower_lsp::jsonrpc::Result;
use tower_lsp::lsp_types::*;
use tower_lsp::Client;
use tower_lsp::LanguageServer;
use tower_lsp::LspService;
use tower_lsp::Server;
use tower_lsp::{jsonrpc, ClientSocket};

use crate::main_loop::Event;
use crate::main_loop::GlobalState;
use crate::main_loop::TokioUnboundedSender;

// Based on https://stackoverflow.com/a/69324393/1725177
macro_rules! cast_response {
    ($target:expr, $pat:path) => {{
        match $target {
            Ok($pat(resp)) => Ok(resp),
            Err(err) => Err(new_jsonrpc_error(format!("{err:?}"))),
            _ => panic!("Unexpected variant while casting to {}", stringify!($pat)),
        }
    }};
}

#[derive(Debug)]
pub(crate) enum LspMessage {
    Notification(LspNotification),
    Request(
        LspRequest,
        TokioUnboundedSender<anyhow::Result<LspResponse>>,
    ),
}

#[derive(Debug)]
pub(crate) enum LspNotification {
    Initialized(InitializedParams),
    DidChangeWorkspaceFolders(DidChangeWorkspaceFoldersParams),
    DidChangeConfiguration(DidChangeConfigurationParams),
    DidChangeWatchedFiles(DidChangeWatchedFilesParams),
    DidOpenTextDocument(DidOpenTextDocumentParams),
    DidChangeTextDocument(DidChangeTextDocumentParams),
    DidSaveTextDocument(DidSaveTextDocumentParams),
    DidCloseTextDocument(DidCloseTextDocumentParams),
}

#[derive(Debug)]
pub(crate) enum LspRequest {
    Initialize(InitializeParams),
    Shutdown(),
}

#[derive(Debug)]
pub(crate) enum LspResponse {
    Initialize(InitializeResult),
    Shutdown(()),
}

#[derive(Debug)]
struct Backend {
    /// Channel for communication with the main loop.
    events_tx: TokioUnboundedSender<Event>,

    /// Handle to main loop. Drop it to cancel the loop, all associated tasks,
    /// and drop all owned state.
    _main_loop: tokio::task::JoinSet<()>,
}

impl Backend {
    async fn request(&self, request: LspRequest) -> anyhow::Result<LspResponse> {
        let (response_tx, mut response_rx) =
            tokio_unbounded_channel::<anyhow::Result<LspResponse>>();

        // Relay request to main loop
        self.events_tx
            .send(Event::Lsp(LspMessage::Request(request, response_tx)))
            .unwrap();

        // Wait for response from main loop
        response_rx.recv().await.unwrap()
    }

    fn notify(&self, notif: LspNotification) {
        // Relay notification to main loop
        self.events_tx
            .send(Event::Lsp(LspMessage::Notification(notif)))
            .unwrap();
    }
}

#[tower_lsp::async_trait]
impl LanguageServer for Backend {
    async fn initialize(&self, params: InitializeParams) -> Result<InitializeResult> {
        cast_response!(
            self.request(LspRequest::Initialize(params)).await,
            LspResponse::Initialize
        )
    }

    async fn initialized(&self, params: InitializedParams) {
        self.notify(LspNotification::Initialized(params));
    }

    async fn shutdown(&self) -> Result<()> {
        cast_response!(
            self.request(LspRequest::Shutdown()).await,
            LspResponse::Shutdown
        )
    }

    async fn did_change_workspace_folders(&self, params: DidChangeWorkspaceFoldersParams) {
        self.notify(LspNotification::DidChangeWorkspaceFolders(params));
    }

    async fn did_change_configuration(&self, params: DidChangeConfigurationParams) {
        self.notify(LspNotification::DidChangeConfiguration(params));
    }

    async fn did_change_watched_files(&self, params: DidChangeWatchedFilesParams) {
        self.notify(LspNotification::DidChangeWatchedFiles(params));
    }

    async fn did_open(&self, params: DidOpenTextDocumentParams) {
        self.notify(LspNotification::DidOpenTextDocument(params));
    }

    async fn did_change(&self, params: DidChangeTextDocumentParams) {
        self.notify(LspNotification::DidChangeTextDocument(params));
    }

    async fn did_save(&self, params: DidSaveTextDocumentParams) {
        self.notify(LspNotification::DidSaveTextDocument(params));
    }

    async fn did_close(&self, params: DidCloseTextDocumentParams) {
        self.notify(LspNotification::DidCloseTextDocument(params));
    }
}

pub async fn start_lsp<I, O>(read: I, write: O)
where
    I: AsyncRead + Unpin,
    O: AsyncWrite,
{
    log::trace!("Starting LSP");

    let (service, socket) = new_lsp();
    let server = Server::new(read, write, socket);
    server.serve(service).await;

    log::trace!("LSP exiting gracefully.",);
}

fn new_lsp() -> (LspService<Backend>, ClientSocket) {
    log::trace!("Starting LSP");

    let init = |client: Client| {
        let state = GlobalState::new(client);
        let events_tx = state.events_tx();

        // Start main loop and hold onto the handle that keeps it alive
        let main_loop = state.start();

        Backend {
            events_tx,
            _main_loop: main_loop,
        }
    };

    LspService::new(init)
}

fn new_jsonrpc_error(message: String) -> jsonrpc::Error {
    jsonrpc::Error {
        code: jsonrpc::ErrorCode::ServerError(-1),
        message: message.into(),
        data: None,
    }
}

#[cfg(test)]
mod tests {
    use futures::{executor::block_on, StreamExt};
    use tokio::io::{ReadHalf, SimplexStream, WriteHalf};
    use tower_lsp::lsp_types;

    use super::*;
    use crate::tests::codec::LanguageServerCodec;
    use crate::tests::request::Request;

    use futures_util::sink::SinkExt;
    use tokio_util::codec::{FramedRead, FramedWrite};

    struct TestClient {
        pub rx: FramedRead<ReadHalf<SimplexStream>, LanguageServerCodec<jsonrpc::Response>>,
        pub tx: FramedWrite<WriteHalf<SimplexStream>, LanguageServerCodec<Request>>,

        server_handle: Option<tokio::task::JoinHandle<()>>,
        id_counter: i64,
    }

    impl TestClient {
        pub fn new() -> Self {
            let (client_rx, mut client_tx) = tokio::io::simplex(1024);
            let (mut server_rx, server_tx) = tokio::io::simplex(1024);

            let server_handle =
                tokio::spawn(async move { start_lsp(&mut server_rx, &mut client_tx).await });

            let rx = FramedRead::new(client_rx, LanguageServerCodec::default());
            let tx = FramedWrite::new(server_tx, LanguageServerCodec::default());

            Self {
                rx,
                tx,
                server_handle: Some(server_handle),
                id_counter: 0,
            }
        }

        // `jsonrpc::Id` requires i64 IDs
        fn id(&mut self) -> i64 {
            let id = self.id_counter;
            self.id_counter = id + 1;
            id
        }

        pub async fn recv_response(&mut self) -> jsonrpc::Response {
            // Unwrap: Option (None if stream closed), then Result (Err if codec fails).
            self.rx.next().await.unwrap().unwrap()
        }

        pub async fn request<R>(&mut self, params: R::Params) -> i64
        where
            R: lsp_types::request::Request,
        {
            let id = self.id();
            let req = Request::from_request::<R>(jsonrpc::Id::Number(id), params);

            // Unwrap: For this test client it's fine to panic if we can't send
            self.tx.send(req).await.unwrap();

            id
        }

        async fn initialize(&mut self) -> i64 {
            let params = lsp_types::InitializeParams::default();
            self.request::<lsp_types::request::Initialize>(params).await
        }

        async fn shutdown(&mut self) -> i64 {
            self.request::<lsp_types::request::Shutdown>(()).await
        }
    }

    impl Drop for TestClient {
        fn drop(&mut self) {
            // TODO: Check that no messages are pending

            // Unwrap: We drop only once, so handle must be Some
            let _handle = std::mem::take(&mut self.server_handle).unwrap();

            block_on(async {
                self.shutdown().await;

                // TODO: Implement Shutdown
                // Unwrap: Panics if task can't shut down as expected
                // handle.await.unwrap();
            })
        }
    }

    #[tokio::test]
    async fn test_init() {
        let mut client = TestClient::new();

        client.initialize().await;

        let value = client.recv_response().await;
        println!("{value:?}");
    }
}
