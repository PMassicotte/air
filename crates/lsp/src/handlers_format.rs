//
// handlers_format.rs
//
// Copyright (C) 2024 Posit Software, PBC. All rights reserved.
//
//

use air_r_formatter::{context::RFormatOptions, format_node};
use air_r_syntax::{RSyntaxKind, RSyntaxNode, WalkEvent};
use biome_formatter::{IndentStyle, LineWidth};
use biome_rowan::{AstNode, Language, SyntaxElement};
use biome_text_size::TextRange;
use tower_lsp::lsp_types;

use crate::state::WorldState;
use crate::{from_proto, to_proto};

#[tracing::instrument(level = "info", skip_all)]
pub(crate) fn document_formatting(
    params: lsp_types::DocumentFormattingParams,
    state: &WorldState,
) -> anyhow::Result<Option<Vec<lsp_types::TextEdit>>> {
    let doc = state.get_document(&params.text_document.uri)?;

    let line_width = LineWidth::try_from(80).map_err(|err| anyhow::anyhow!("{err}"))?;

    // TODO: Handle FormattingOptions
    let options = RFormatOptions::default()
        .with_indent_style(IndentStyle::Space)
        .with_line_width(line_width);

    if doc.parse.has_errors() {
        return Err(anyhow::anyhow!("Can't format when there are parse errors."));
    }

    let formatted = format_node(options.clone(), &doc.parse.syntax())?;
    let output = formatted.print()?.into_code();

    // Do we need to check that `doc` is indeed an R file? What about special
    // files that don't have extensions like `NAMESPACE`, do we hard-code a
    // list? What about unnamed temporary files?

    let edits = to_proto::replace_all_edit(&doc.line_index, &doc.contents, &output)?;
    Ok(Some(edits))
}

#[tracing::instrument(level = "info", skip_all)]
pub(crate) fn document_range_formatting(
    params: lsp_types::DocumentRangeFormattingParams,
    state: &WorldState,
) -> anyhow::Result<Option<Vec<lsp_types::TextEdit>>> {
    let doc = state.get_document(&params.text_document.uri)?;

    let line_width = LineWidth::try_from(80).map_err(|err| anyhow::anyhow!("{err}"))?;
    let range =
        from_proto::text_range(&doc.line_index.index, params.range, doc.line_index.encoding)?;

    // TODO: Handle FormattingOptions
    let options = RFormatOptions::default()
        .with_indent_style(IndentStyle::Space)
        .with_line_width(line_width);

    let logical_lines = find_deepest_enclosing_logical_lines(doc.parse.syntax(), range);
    if logical_lines.is_empty() {
        tracing::warn!("Can't find logical line");
        return Ok(None);
    };

    // Find the overall formatting range by concatenating the ranges of the logical lines.
    // We use the "non-whitespace-range" as that corresponds to what Biome will format.
    let format_range = logical_lines
        .iter()
        .map(|line| text_non_whitespace_range(line))
        .reduce(|acc, new| acc.cover(new))
        .expect("`logical_lines` is non-empty");

    // We need to wrap in an `RRoot` otherwise the comments get attached too
    // deep in the tree. See `CommentsBuilderVisitor` in biome_formatter and the
    // `is_root` logic. Note that `node` needs to be wrapped in at least two
    // other nodes in order to fix this problem, and here we have an `RRoot` and
    // `RExpressionList` that do the job.
    //
    // Since we only format logical lines, it is fine to wrap in an expression list.
    let Some(exprs): Option<Vec<air_r_syntax::AnyRExpression>> = logical_lines
        .into_iter()
        .map(|node| air_r_syntax::AnyRExpression::cast(node))
        .collect()
    else {
        tracing::warn!("Can't cast to `AnyRExpression`");
        return Ok(None);
    };

    let list = air_r_factory::r_expression_list(exprs);
    let eof = air_r_syntax::RSyntaxToken::new_detached(RSyntaxKind::EOF, "", vec![], vec![]);
    let root = air_r_factory::r_root(list, eof).build();

    let format_info = biome_formatter::format_sub_tree(
        root.syntax(),
        air_r_formatter::RFormatLanguage::new(options),
    )?;

    if format_info.range().is_none() {
        // Happens in edge cases when biome returns a `Printed::new_empty()`
        return Ok(None);
    };

    let mut format_text = format_info.into_code();

    // Remove last hard break line from our artifical expression list
    format_text.pop();
    let edits = to_proto::replace_range_edit(&doc.line_index, format_range, format_text)?;

    Ok(Some(edits))
}

// From biome_formatter
fn text_non_whitespace_range<E, L>(elem: &E) -> TextRange
where
    E: Into<SyntaxElement<L>> + Clone,
    L: Language,
{
    let elem: SyntaxElement<L> = elem.clone().into();

    let start = elem
        .leading_trivia()
        .into_iter()
        .flat_map(|trivia| trivia.pieces())
        .find_map(|piece| {
            if piece.is_whitespace() || piece.is_newline() {
                None
            } else {
                Some(piece.text_range().start())
            }
        })
        .unwrap_or_else(|| elem.text_trimmed_range().start());

    let end = elem
        .trailing_trivia()
        .into_iter()
        .flat_map(|trivia| trivia.pieces().rev())
        .find_map(|piece| {
            if piece.is_whitespace() || piece.is_newline() {
                None
            } else {
                Some(piece.text_range().end())
            }
        })
        .unwrap_or_else(|| elem.text_trimmed_range().end());

    TextRange::new(start, end)
}

/// Finds consecutive logical lines. Currently that's only expressions at
/// top-level or in a braced list.
fn find_deepest_enclosing_logical_lines(node: RSyntaxNode, range: TextRange) -> Vec<RSyntaxNode> {
    let mut preorder = node.preorder();
    let mut logical_lines: Vec<RSyntaxNode> = vec![];

    while let Some(event) = preorder.next() {
        match event {
            WalkEvent::Enter(node) => {
                let Some(parent) = node.parent() else {
                    continue;
                };

                let node_range = node.text_trimmed_range();
                if !range.contains_range(node_range) {
                    continue;
                }

                if parent.kind() == RSyntaxKind::R_EXPRESSION_LIST {
                    logical_lines.push(node.clone());
                    preorder.skip_subtree();
                    continue;
                }
            }

            WalkEvent::Leave(_) => {}
        }
    }

    logical_lines
}

#[cfg(test)]
mod tests {
    use biome_text_size::{TextRange, TextSize};

    use crate::{
        documents::Document, tower_lsp::init_test_client, tower_lsp_test_client::TestClientExt,
    };

    #[tests_macros::lsp_test]
    async fn test_format() {
        let mut client = init_test_client().await;

        #[rustfmt::skip]
        let doc = Document::doodle(
"
1
2+2
3 + 3 +
3",
        );

        let formatted = client.format_document(&doc).await;
        insta::assert_snapshot!(formatted);

        client
    }

    // https://github.com/posit-dev/air/issues/61
    #[tests_macros::lsp_test]
    async fn test_format_minimal_diff() {
        let mut client = init_test_client().await;

        #[rustfmt::skip]
        let doc = Document::doodle(
"1
2+2
3
",
        );

        let edits = client.format_document_edits(&doc).await.unwrap();
        assert!(edits.len() == 1);

        let edit = &edits[0];
        assert_eq!(edit.new_text, " + ");

        client
    }

    #[tests_macros::lsp_test]
    async fn test_format_range_none() {
        let mut client = init_test_client().await;

        #[rustfmt::skip]
        let doc = Document::doodle(
"",
        );
        let range = TextRange::new(TextSize::from(0), TextSize::from(0));

        let output = client.format_document_range(&doc, range).await;
        insta::assert_snapshot!(output);

        #[rustfmt::skip]
        let doc = Document::doodle(
"
",
        );
        let range = TextRange::new(TextSize::from(0), TextSize::from(1));

        let output = client.format_document_range(&doc, range).await;
        insta::assert_snapshot!(output);

        #[rustfmt::skip]
        let doc = Document::doodle(
"1
",
        );
        let range = TextRange::new(TextSize::from(0), TextSize::from(1));

        let output = client.format_document_range(&doc, range).await;
        insta::assert_snapshot!(output);

        client
    }

    #[tests_macros::lsp_test]
    async fn test_format_range_logical_lines() {
        let mut client = init_test_client().await;

        // 2+2 is the logical line to format
        #[rustfmt::skip]
        let doc = Document::doodle(
"1+1
2+2
",
        );
        let range = TextRange::new(TextSize::from(4), TextSize::from(7));
        let output = client.format_document_range(&doc, range).await;
        insta::assert_snapshot!(output);

        #[rustfmt::skip]
        let doc = Document::doodle(
"1+1
#
2+2
",
        );
        let range = TextRange::new(TextSize::from(6), TextSize::from(9));

        let output = client.format_document_range(&doc, range).await;
        insta::assert_snapshot!(output);

        // The element in the braced expression is a logical line
        // FIXME: Should this be the whole `{2+2}` instead?
        #[rustfmt::skip]
        let doc = Document::doodle(
"1+1
{2+2}
",
        );

        let range = TextRange::new(TextSize::from(5), TextSize::from(8));
        let output = client.format_document_range(&doc, range).await;
        insta::assert_snapshot!(output);

        // Including braces
        let range = TextRange::new(TextSize::from(4), TextSize::from(9));
        let output = client.format_document_range(&doc, range).await;
        insta::assert_snapshot!(output);

        // The deepest element in the braced expression is our target
        #[rustfmt::skip]
        let doc = Document::doodle(
"1+1
{
  2+2
  {
    3+3
  }
}
",
        );

        let range = TextRange::new(TextSize::from(20), TextSize::from(23));
        let output = client.format_document_range(&doc, range).await;
        insta::assert_snapshot!(output);
        client
    }

    #[tests_macros::lsp_test]
    async fn test_format_range_mismatched_indent() {
        let mut client = init_test_client().await;

        #[rustfmt::skip]
        let doc = Document::doodle(
"1
  2+2
",
        );

        // We don't change indentation when `2+2` is formatted
        let range = TextRange::new(TextSize::from(4), TextSize::from(7));
        let output = client.format_document_range(&doc, range).await;
        insta::assert_snapshot!(output);

        // Debatable: Should we make an effort to remove unneeded indentation
        // when it's part of the range?
        let range_wide = TextRange::new(TextSize::from(2), TextSize::from(7));
        let output_wide = client.format_document_range(&doc, range_wide).await;
        assert_eq!(output, output_wide);

        client
    }

    #[tests_macros::lsp_test]
    async fn test_format_range_multiple_lines() {
        let mut client = init_test_client().await;

        #[rustfmt::skip]
        let doc = Document::doodle(
"1+1
#
2+2
",
        );

        // Selecting the last two lines
        let range = TextRange::new(TextSize::from(4), TextSize::from(9));
        let output = client.format_document_range(&doc, range).await;
        insta::assert_snapshot!(output);

        // Selecting all three lines
        let range = TextRange::new(TextSize::from(0), TextSize::from(9));
        let output = client.format_document_range(&doc, range).await;
        insta::assert_snapshot!(output);

        client
    }
}
