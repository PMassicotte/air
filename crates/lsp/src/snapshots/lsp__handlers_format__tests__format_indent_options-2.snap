---
source: crates/lsp/src/handlers_format.rs
expression: output_tab
---
{
	1
}
